packages/material-ui/src/BusyButton/BusyButton.js

```diff
@@ -0,0 +1,144 @@
+  pendingIndicatorStart: {
+    left: 16,
+  },
+  /* Styles applied to the pendingIndicator element if `pendingPosition="end"`. */
+  pendingIndicatorEnd: {
+    right: 16,
+  },
+  /* Styles applied to the endIcon element if `pending={true}` and `pendingPosition="end"`. */
+  endIconLoadingEnd: {
+    visibility: 'hidden',
+  },
+  /* Styles applied to the startIcon element if `pending={true}` and `pendingPosition="start"`. */
+  startIconLoadingStart: {
+    visibility: 'hidden',
+  },
+  /* Styles applied to the label element if `pending={true}` and `pendingPosition="center"`. */
+  labelLoadingCenter: {
+    visibility: 'hidden',
+  },
+});
+
+const PendingIndicator = <CircularProgress color="inherit" size={16} />;
+
+const BusyButton = React.forwardRef(function BusyButton(props, ref) {
+  const {
+    children,
+    classes,
+    className,
+    disabled = false,
+    pending = false,
+    pendingIndicator = PendingIndicator,
+    pendingPosition = 'center',
+    ...other
+  } = props;
+
+  return (
+    <Button
+      className={clsx(
+        classes.root,
+        {
+          [classes.pending]: pending,
+        },
+        className,
+      )}
+      disabled={disabled || pending}
+      ref={ref}
+      classes={{
+        startIcon: classes[`startIcon${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+        endIcon: classes[`endIcon${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+        label: classes[`label${pending ? 'Loading' : ''}${capitalize(pendingPosition)}`],
+      }}
+      {...other}
+    >
+      {pending && (
+        <div
+          className={clsx(
+            classes.pendingIndicator,
+            classes[`pendingIndicator${capitalize(pendingPosition)}`],
+          )}
+        >
+          {pendingIndicator}
+        </div>
+      )}
+
+      {children}
+    </Button>
+  );
+});
+
+BusyButton.propTypes = {
+  // ----------------------------- Warning --------------------------------
+  // | These PropTypes are generated from the TypeScript type definitions |
+  // |     To update them edit the d.ts file and run "yarn proptypes"     |
+  // ----------------------------------------------------------------------
+  /**
+   * The content of the button.
+   */
+  children: PropTypes.node,
+  /**
+   * Override or extend the styles applied to the component.
+   * See [CSS API](#css) below for more details.
+   */
+  classes: PropTypes.object,
+  /**
+   * @ignore
+   */
+  className: PropTypes.string,
+  /**
+   * If `true`, the button will be disabled.
+   */
+  disabled: PropTypes.bool,
+  /**
+   * The content of the button.
+   */
+  pending: PropTypes.bool,
+  /**
+   * Element placed before the children if the button is in pending state.
+   */
+  pendingIndicator: PropTypes.node,
+  /**
+   * The pending indicator can be positioned on the start, end, or the center of the button.
+   */
+  pendingPosition: chainPropTypes(PropTypes.oneOf(['start', 'end', 'center']), (props) => {
```

Comment:
Have we considered this option: the BusyButton set the `pendingPosition` depending on the props. If `startIcon` is provided and `pendingPosition`, is not it sets `pendingPosition="start"` automatically?

Reply:
I think that make sense 👍 (+ defaulting to `pendingPosition="end"` when there is `endIcon`. May be unpredictable when there are both start and end icon, but not sure how common that is...). But anyway I still think we will need validations for the props (in case if someone specifies `pendingPosition="start"` without `startIcon` and respectively for the end combination)

What do you think?

Reply:
I have no preference, happy either way

Reply:
To be honest I like the `center` as a default value because it works always without surprises. Then if people see that that is not what they want, they can opt out by adding `pendingPosition` to `start`/`end`. So would suggest keeping it for now as this, and we can update it later if we see people complaining..

Reply:
Sure, let's do that 👍